nixos/modules/virtualisation/waagent.nix

```
@@ -0,0 +1,364 @@
+          '';
+        };
+
+        EnableSwap = mkEnableOption ''
+          If enabled, the agent creates a swap file (`/swapfile`) on the resource disk
+          and adds it to the system swap space.
+
+          This configuration has no effect if resource disk is managed by cloud-init.
+        '';
+
+        SwapSizeMB = mkOption {
+          type = types.int;
+          default = 0;
+          description = ''
+            Specifies the size of the swap file in megabytes.
+
+            This configuration has no effect if resource disk is managed by cloud-init.
+          '';
+        };
+      };
+
+      Logs.Verbose = lib.mkEnableOption ''
+        If you set this option, log verbosity is boosted.
+        Waagent logs to `/var/log/waagent.log` and uses the system logrotate functionality to rotate logs.
+      '';
+
+      OS = {
+        EnableRDMA = lib.mkEnableOption ''
+          If enabled, the agent attempts to install and then load an RDMA kernel driver
+          that matches the version of the firmware on the underlying hardware.
+        '';
+
+        RootDeviceScsiTimeout = lib.mkOption {
+          type = types.nullOr types.int;
+          default = 300;
+          description = ''
+            Configures the SCSI timeout in seconds on the OS disk and data drives.
+            If set to `null`, the system defaults are used.
+          '';
+        };
+      };
+
+      HttpProxy = {
+        Host = lib.mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          description = ''
+            If you set http proxy, waagent will use is proxy to access the Internet.
+          '';
+        };
+
+        Port = lib.mkOption {
+          type = types.nullOr types.int;
+          default = null;
+          description = ''
+            If you set http proxy, waagent will use this proxy to access the Internet.
+          '';
+        };
+      };
+
+      AutoUpdate.Enable = lib.mkEnableOption ''
+        Enable or disable autoupdate for goal state processing.
+      '';
+    };
+  };
+in
+{
+  options.services.waagent = {
+    enable = lib.mkEnableOption ''
+      Whether to enable the Windows Azure Linux Agent.
+    '';
+
+    package = lib.mkPackageOption pkgs "waagent" { };
+
+    extraPackages = lib.mkOption {
+      default = [ ];
+      description = ''
+        Additional packages to add to the waagent {env}`PATH`.
+      '';
+      example = lib.literalExpression "[ pkgs.powershell ]";
+      type = lib.types.listOf lib.types.package;
+    };
+
+    settings = lib.mkOption {
+      type = settingsType;
+      default = { };
+      description = ''
+        The waagent.conf configuration, see https://learn.microsoft.com/en-us/azure/virtual-machines/extensions/agent-linux for documentation.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.settings.HttpProxy.Host == null || cfg.settings.HttpProxy.Port != null;
```

```suggestion
        assertion = cfg.settings.HttpProxy.Host -> cfg.settings.HttpProxy.Port;
```
A logical implication would be better here